lib: posix: nanosleep

cfriedt opened the pull request:
This change adds support for nanosleep(2) documented at
https://pubs.opengroup.org/onlinepubs/9699919799/functions/nanosleep.html

Note: Does not achieve any greater resolution than that provided by `k_busy_wait(us)` but at least makes the API available.

I've not researched it much in Zephyr, but I would say there is still some work to be done for a generic clock API - one that allows querying resolutions, supports multiple system clocks, etc, primarily to support POSIX time APIs.

A nanosleep system call would also require high-res timer support (See #6498)

Fixes #25554

pfalcon commented on lib/posix/sleep.c:
```diff
@@ -5,14 +5,17 @@
  * See IEEE 1003.1
  */
-unsigned sleep(unsigned int seconds)
+unsigned int sleep(unsigned int seconds)
```
It's good zen exercise to try to avoid patching unrelated (== not directly related) code ;-).

cfriedt replied:
Will fix. That was a checkpatch.pl / fix-inline modification.

cfriedt committed:
lib: posix: nanosleep

This change adds support for nanosleep(2) documented at
https://pubs.opengroup.org/onlinepubs/9699919799/functions/nanosleep.html

N.B: Currently, this provides no better resolution than
k_busy_wait()

Fixes #25554

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
I have a feeling there is a conflict with `CONFIG_BOARD_NATIVE_POSIX`
`CONFIG_BOARD_NATIVE_POSIX_64BIT`

pfalcon commented:
> I have a feeling there is a conflict with CONFIG_BOARD_NATIVE_POSIX
CONFIG_BOARD_NATIVE_POSIX_64BIT

I'm not sure what exactly you mean (that native_posix and native_posix_64 boards conflict between each other?), but otherwise it's a known problem that Zephyr POSIX subsys conflicts with native_posix board. It should be understanding why - it's too layery a pie, where Zephyr POSIX subsys sits on top of native Zephyr core, and naive_posix boards tries to implement that in terms of the underlying full-POSIX system. Something doesn't match regularly. Many a POSIX tests/samples simply blacklist native_posix.

cfriedt commented:
> > I have a feeling there is a conflict with CONFIG_BOARD_NATIVE_POSIX
> > CONFIG_BOARD_NATIVE_POSIX_64BIT
> 
> I'm not sure what exactly you mean (that native_posix and native_posix_64 boards conflict between each other?), but otherwise it's a known problem that Zephyr POSIX subsys conflicts with native_posix board. It should be understanding why - it's too layery a pie, where Zephyr POSIX subsys sits on top 

That's what I meant. Sorry, was on mobile when I commented previously.

cfriedt commented:
Should be good to go now
```
$ ./scripts/sanitycheck --enable-slow --inline-logs -j8 --testcase-root tests/posix/common
Renaming output directory to /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out.24
INFO    - JOBS: 8
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 26 test configurations selected, 972 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   26/  26  100%  skipped:    0, failed:    0
INFO    - 26 of 26 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 179.42 seconds
INFO    - In total 4 test cases were executed on 13 out of total 250 platforms (5.20%)
INFO    - 20 tests executed on platforms, 6 tests were only built.
```

cfriedt commented:
Btw, the variance in latency (in CI / qemu) is pretty drastic between architectures. E.g. on qemu_cortex_r5, it was 500us, on xtensa it was 86 us, and on qemu_cortex_m3 it was 25us.

Sample size of 1, not statistically representative, of course.

cfriedt commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -0,0 +1,214 @@
+	ns += s * NS_PER_SEC;
+
+	dt = now - then;
+	dt_ns = k_cyc_to_ns_floor64(dt);
```
Looks like I need the ceil variant here. Although ns (realistically??)

pfalcon commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	if (rqtp->tv_sec == 0 && rqtp->tv_nsec == 0) {
+		goto maybe_write;
```
Label name sounds surprising. It's rather "return_val" or "skip" to me. Probably not worth changing, just thinking aloud.

pfalcon reviewed (approved):
Thanks!

pfalcon commented:
@cfriedt, Thanks for the patch. I would personally recommend this merged for 2.3, but per the release freeze rules, 2 reviews are required at this time. I've added @carlescufi who both could review and merged (but definitely busy with a lot of other stuff now, as the release manager for this version), but feel free ping other folks who you know can help with review.

cfriedt commented:
> @cfriedt, Thanks for the patch. I would personally recommend this merged for 2.3, but per the release freeze rules, 2 reviews are required at this time. I've added @carlescufi who both could review and merged (but definitely busy with a lot of other stuff now, as the release manager for this version), but feel free ping other folks who you know can help with review.

It would be awesome if nanosleep could actually get relatively close to the exact time (if there is hardware to support it). The high-res timer stuff is pretty important though, and I think it deserves a great deal of care. No rush on that one. The nice thing about it though is that the interface is there, and it's not moving, so just redoing the implementation later will have no unwanted side-effects.

jukkar commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+/* required for struct timespec */
+#include <posix/time.h>
+
+#define NS_PER_SEC 1000000000ULL
```
There is `NSEC_PER_SEC` defined in `include/sys_clock.h` so we should use that

jukkar commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+#include <posix/time.h>
+
+#define NS_PER_SEC 1000000000ULL
+#define NS_PER_USEC 1000ULL
```
Also `NSEC_PER_USEC` is also defined so lets use that instead

jukkar commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+int nanosleep(const struct timespec *rqtp, struct timespec *rmtp)
+{
+	u64_t ns;
+	const bool write = (rmtp != NULL);
```
Better would be something like `do_write` so that when grepping for `write` function we would not get this one.
Also () are superfluous and could be removed, but if you think it improves readability then I am fine with this too.

jukkar commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+	if (rqtp == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
```
please place empty lines after } ending a block

jukkar commented on lib/posix/nanosleep.c:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	/* currently we have no mechanism to achieve greater resolution */
+	k_busy_wait(ns / NS_PER_USEC);
```
Hmm, could we still call `k_sleep()` as it supports us sleep values, not sure if it can actually sleep that short time.

pfalcon replied:
I assume the usecase for nanosleep() is really short intervals, which also should be as precise as possible, that's why @cfriedt used k_busy_wait(). k_sleep() leads to rescheduling of the thread, so any hopes for precision are lost.

jukkar commented on tests/posix/common/src/nanosleep.c:
```diff
@@ -0,0 +1,221 @@
+#include <errno.h>
+#include <posix/time.h>
+
+#define NS_PER_SEC 1000000000ULL
```
We should use here `NSEC_PER_SEC`

cfriedt replied:
Sorry, I thought I had fixed that already, thanks for pointing it out.

jukkar reviewed (approved):
Looks good, minor nit about the macro name that was missed from previous cleanup.

cfriedt committed:
tests: posix: tests for nanosleep(2)

This commit provides test-cases for nanosleep(2)

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
@carlescufi  - rebased off of using `<stdint.h>` integer types. Should be good to go as long as the build works.

cfriedt commented:
Was green before. Looks like there is some misra change that breaks things atm

pfalcon commented:
> some misra change

New swear word in software engineering? ;-)

pfalcon commented:
~~~
starting test - test_timeout_lifo_thread

    Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/lifo/lifo_usage/src/main.c:329: test_timeout_lifo_thread: packet != NULL is false

FAIL - test_timeout_lifo_thread
~~~

I got such a failure in one of my recent PR. Looks like nondeterministic failure du jour.

@cfriedt: Please keep rebasing this, until sweet spot is found.
